utils/initial_data_preparation.py

```diff
@@ -67,22 +75,123 @@ def parse_args():
 
 
 def check_and_clean_data(df):
+    """
+    Processes the provided DataFrame by removing rows containing null values and
+    duplicate rows, updating the DataFrame in place.
+
+    """
     # Checking for null values
     null_counts = df.isnull().sum()
     if np.any(null_counts):
         logger.info("Null values found:")
         logger.info(null_counts[null_counts > 0])
+
+        # Count rows before dropping
+        initial_count = len(df)
+
+        # Drop rows with any null values in place
+        df.dropna(inplace=True)
+
+        # Log how many rows were dropped
+        rows_dropped = initial_count - len(df)
+        logger.info(f"Dropped {rows_dropped} rows containing null values.")
     else:
         logger.info("No null values found.")
 
     # Checking for duplicates
     duplicates = df.duplicated().sum()
     if duplicates:
         logger.info(f"Number of duplicate rows: {duplicates}")
+        # I will keep the first occurrence of each duplicate row and remove the
+        # others:
+        df = df.drop_duplicates(keep="first")
+        logger.info("First occurrence of each duplicate row has been kept. ")
     else:
         logger.info("No duplicate rows found.")
 
 
+def filter_out_incomplete_urls(df):
+    """
+    Filters rows in a DataFrame based on the completeness of the 'repourl'
+    URLs. A URL is considered complete if it contains at least five parts,
+    including the protocol, empty segment (for '//'), domain, and at least
+    two path segments, e.g., 'https://github.com/owner/repo'. Raises
+    an error if the required column 'repourl' is missing.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing URLs.
+
+    Returns:
+        pd.DataFrame: DataFrame with rows containing complete URLs.
+
+    Raises:
+        ValueError: If the 'repourl' column is missing from the DataFrame.
+    """
+    if "repourl" not in df.columns:
+        logger.critical(
+            "Critical: DataFrame columns are: {}".format(df.columns.tolist())
+        )
+        logger.critical(
+            "DataFrame does not contain 'repourl' column. Aborting operation."
+        )
+        raise ValueError("DataFrame must contain a 'repourl' column.")
+
+    # Helper function to determine if a URL is complete
+    def is_complete_url(url):
+        # Check if the url is not a string
+        if not isinstance(url, str):
+            return False
+        parts = url.rstrip("/").split("/")
```

Comment:
Before this line; add an example url so that it's easy to see why there should be a minimum of 5 parts.

e.g. `https://github.com/owner/repo`

BTW: Presumably a malformed URL such as: `https:/232u9q2/github.com/owner/repo` would also be considered a complete URL. No need to address this in this PR; something we can consider in future is how to detect this form of malformed URL.

Also BTW there are also valid URLs that include a username and password, https://stackoverflow.com/questions/10050877/url-username-with provides a good discussion on this.

Reply:
- Added the example url. 
- The example also exists in the `docstring` of the function.